lib/std/start.zig

```diff
@@ -492,44 +503,47 @@ fn callMainWithArgs(argc: usize, argv: [*][*:0]u8, envp: [][*:0]u8) u8 {
     return initEventLoopAndCallMain();
 }
 
-fn main(c_argc: c_int, c_argv: [*][*:0]c_char, c_envp: [*:null]?[*:0]c_char) callconv(.C) c_int {
+fn cMain(c_argc: c_int, c_argv: [*][*:0]c_char, c_envp: [*:null]?[*:0]c_char) callconv(.C) c_int {
     var env_count: usize = 0;
     while (c_envp[env_count] != null) : (env_count += 1) {}
     const envp = @ptrCast([*][*:0]u8, c_envp)[0..env_count];
 
-    if (builtin.os.tag == .linux) {
+    if (native_os == .linux) {
         const at_phdr = std.c.getauxval(elf.AT_PHDR);
         const at_phnum = std.c.getauxval(elf.AT_PHNUM);
         const phdrs = (@intToPtr([*]elf.Phdr, at_phdr))[0..at_phnum];
-        expandStackSize(phdrs);
+        linuxExpandStackSize(phdrs);
     }
 
-    return @call(.always_inline, callMainWithArgs, .{ @intCast(usize, c_argc), @ptrCast([*][*:0]u8, c_argv), envp });
+    return @call(.always_inline, posixCallMainWithArgs, .{ @intCast(usize, c_argc), @ptrCast([*][*:0]u8, c_argv), envp });
 }
 
-fn mainWithoutEnv(c_argc: c_int, c_argv: [*][*:0]c_char) callconv(.C) c_int {
+fn cMainWithoutEnv(c_argc: c_int, c_argv: [*][*:0]c_char) callconv(.C) c_int {
     std.os.argv = @ptrCast([*][*:0]u8, c_argv)[0..@intCast(usize, c_argc)];
     return @call(.always_inline, callMain, .{});
 }
 
-// General error message for a malformed return type
-const bad_main_ret = "expected return type of main to be 'void', '!void', 'noreturn', 'u8', or '!u8'";
+/// Error message for a malformed main return type.
+const bad_main_ret = if (std.os.ExitStatus == void)
+    "expected return type of main to be 'void', '!void, or 'noreturn'"
+else
+    std.fmt.comptimePrint("expected return type of main to be 'void', '!void, '{[ty]s}', '!{[ty]s}', or 'noreturn'", .{ .ty = @typeName(std.os.ExitStatus) });
 
 // This is marked inline because for some reason LLVM in release mode fails to inline it,
 // and we want fewer call frames in stack traces.
-inline fn initEventLoopAndCallMain() u8 {
+inline fn initEventLoopAndCallMain() CallMainReturnType {
     if (std.event.Loop.instance) |loop| {
         if (loop == std.event.Loop.default_instance) {
             loop.init() catch |err| {
                 std.log.err("{s}", .{@errorName(err)});
```

Comment:
@fifty-six actually, we can refactor and do the whole `outputString` thing here too. In theory it is possible to implement the event loop for the UEFI too, right?

Reply:
I *think* you could? There's `EFI_EVENT` and such for non-blocking IO and you can create events as wanted - though I haven't used non-blocking IO with UEFI myself.